app/utils.py

```diff
@@ -68,18 +85,39 @@ async def _make_request(request,
                      method=method,
                      url=url,
                      handler=func.__name__)
+
+        attempt_number = View._make_request.retry.statistics['attempt_number']
         try:
-            async with request.app.http_session_pool.request(
-                    method, url, auth=auth, json=json, ssl=False) as resp:
-                func(resp)
-                if return_json:
-                    return await resp.json()
-                else:
-                    return None
+            if attempt_number > 1:
+                wait_secs = int(request.app['WAIT_BEFORE_RETRY'])
+                time.sleep(wait_secs)
```

Comment:
I'm a bit concerned at the 2 second sleep before the first retry. It's a pretty long time and I think that it alone is higher than our NFR response times. 

Until the keep-alive expires after 5 seconds then I assume that the typical path for a connection talking to a recently departed RHSVC is:
1. Send request to just departed RHSVC (this fails).
2. Wait for 2 seconds.
3. Retry with basic connection. Virtually always successful.

As the first attempt with the basic connection is probably going to succeed couldn't it only do a sleep on the 2nd and subsequent attempt (ie, attempt no.3 and onwards) ?

Reply:
In practice, without the sleep I found that sometimes the retry was not successful. The wait solved a lot of problems. 

Since the wait only ever happens a handful of times on scale-back, then it really is "once in a blue moon" and I thought it was acceptable given that this will happen so infrequently.

My rational for 2 seconds was that in the pathological case 2 seconds x 5 attempts would mean the terminated pod has completely disappeared, and once that happens, I think we won't have anymore connection fails for that pod.

The NFR is surely not applicable for every single connection. Its an average right ?

Reply:
I've discussed this with Pete, and he is keen to optimise the retry wait case, so I will investigate a lighter weight approach (probably escalating wait time).

Reply:
I have implemented a rising scale of wait times now. The configuration parameter is no longer a time in seconds, but is an exponent which is applied to the (attempt-number - 1).

By default, the first retry will have 0 delay, the second retry 1 second, then 2 seconds , 3 seconds ... until the attempts give up.